### PR TITLE
Simplify Property code

### PR DESCRIPTION
**Pull request by WeirdRubberDuck**
- Remove some unused property types. 
- Remove `PropertyDelegate`
- Write helper function for converting to Lua / JSON

### CODE REVIEW

File: src/properties/list/intlistproperty.cpp [hunk 1]
```diff
@@ -59,46 +72,21 @@ std::vector<int> fromLuaConversion(lua_State* state, bool& success) {
...
 
-bool toLuaConversion(lua_State* state, std::vector<int> val) {
-    lua_createtable(state, static_cast<int>(val.size()), 0);
+void IntListProperty::toLuaConversion(lua_State* state) const {
+    lua_createtable(state, static_cast<int>(_value.size()), 0);
 
     int i = 1;
-    for (const int& v : val) {
+    for (int v : _value) {
         lua_pushinteger(state, i);
```
Comment (alexanderbock) [inline src/properties/list/intlistproperty.cpp:80]: ^^

File: src/properties/list/doublelistproperty.cpp [hunk 2]
```diff
@@ -59,46 +73,21 @@ std::vector<double> fromLuaConversion(lua_State* state, bool& success) {
...
 
-bool toLuaConversion(lua_State* state, std::vector<double> val) {
-    lua_createtable(state, static_cast<int>(val.size()), 0);
+void DoubleListProperty::toLuaConversion(lua_State* state) const {
+    lua_createtable(state, static_cast<int>(_value.size()), 0);
 
     int i = 1;
-    for (int v : val) {
+    for (double v : _value) {
         lua_pushinteger(state, i);
```
[the remark this answers is not in the document]
Comment (alexanderbock) [reply src/properties/list/doublelistproperty.cpp:81]: Thinking about it, might it make sense to have a general implementation that just does `ghoul::lua::push` in the NumericalProperty and only overwrite it for the types that need to do something more than just push a single value?

File: src/properties/scalar/boolproperty.cpp [hunk 3]
```diff
@@ -23,46 +23,33 @@
...
+bool BoolProperty::fromLuaConversion(lua_State* state, bool& success) const {
+    success = (lua_isboolean(state, -1) == 1);
+    return success ? (lua_toboolean(state, -1) == 1) : false;
 }
 
-} // namespace
-
-namespace openspace::properties {
+void BoolProperty::toLuaConversion(lua_State* state) const {
+    lua_pushboolean(state, _value);
```
Comment (alexanderbock) [inline src/properties/scalar/boolproperty.cpp:48]: ^^

File: src/properties/scalar/doubleproperty.cpp [hunk 4]
```diff
@@ -42,31 +51,8 @@ double fromLuaConversion(lua_State* state, bool& success) {
     }
 }
 
-bool toLuaConversion(lua_State* state, double value) {
-    lua_pushnumber(state, value);
-    return true;
+void DoubleProperty::toLuaConversion(lua_State* state) const {
+    lua_pushnumber(state, _value);
```
Comment (alexanderbock) [inline src/properties/scalar/doubleproperty.cpp:55]: ^^

File: src/properties/scalar/floatproperty.cpp [hunk 5]
```diff
@@ -42,31 +51,8 @@ float fromLuaConversion(lua_State* state, bool& success) {
     }
 }
 
-bool toLuaConversion(lua_State* state, float value) {
-    lua_pushnumber(state, static_cast<lua_Number>(value));
-    return true;
+void FloatProperty::toLuaConversion(lua_State* state) const {
+    lua_pushnumber(state, static_cast<lua_Number>(_value));
```
Comment (alexanderbock) [inline src/properties/scalar/floatproperty.cpp:55]: ^^

File: src/properties/scalar/intproperty.cpp [hunk 6]
```diff
@@ -42,31 +51,8 @@ int fromLuaConversion(lua_State* state, bool& success) {
     }
 }
 
-bool toLuaConversion(lua_State* state, int value) {
-    lua_pushnumber(state, static_cast<lua_Number>(value));
-    return true;
+void IntProperty::toLuaConversion(lua_State* state) const {
+    lua_pushnumber(state, static_cast<lua_Number>(_value));
```
Comment (alexanderbock) [inline src/properties/scalar/intproperty.cpp:55]: ^^

File: src/properties/scalar/longproperty.cpp [hunk 7]
```diff
@@ -42,31 +51,8 @@ long fromLuaConversion(lua_State* state, bool& success) {
     }
 }
 
-bool toLuaConversion(lua_State* state, long value) {
-    lua_pushnumber(state, static_cast<lua_Number>(value));
-    return true;
+void LongProperty::toLuaConversion(lua_State* state) const {
+    lua_pushnumber(state, static_cast<lua_Number>(_value));
```
Comment (alexanderbock) [inline src/properties/scalar/longproperty.cpp:55]: ^^

File: src/properties/scalar/shortproperty.cpp [hunk 8]
```diff
@@ -42,31 +51,8 @@ short fromLuaConversion(lua_State* state, bool& success) {
     }
 }
 
-bool toLuaConversion(lua_State* state, short value) {
-    lua_pushnumber(state, static_cast<lua_Number>(value));
-    return true;
+void ShortProperty::toLuaConversion(lua_State* state) const {
+    lua_pushnumber(state, static_cast<lua_Number>(_value));
```
Comment (alexanderbock) [inline src/properties/scalar/shortproperty.cpp:55]: ^^

File: src/properties/scalar/uintproperty.cpp [hunk 9]
```diff
@@ -42,31 +58,8 @@ unsigned int fromLuaConversion(lua_State* state, bool& success) {
     }
 }
 
-bool toLuaConversion(lua_State* state, unsigned int value) {
-    lua_pushnumber(state, static_cast<lua_Number>(value));
-    return true;
+void UIntProperty::toLuaConversion(lua_State* state) const {
+    lua_pushnumber(state, static_cast<lua_Number>(_value));
```
Comment (alexanderbock) [inline src/properties/scalar/uintproperty.cpp:62]: ^^

File: src/properties/scalar/ulongproperty.cpp [hunk 10]
```diff
@@ -42,31 +58,8 @@ unsigned long fromLuaConversion(lua_State* state, bool& success) {
     }
 }
 
-bool toLuaConversion(lua_State* state, unsigned long value) {
-    lua_pushnumber(state, static_cast<lua_Number>(value));
-    return true;
+void ULongProperty::toLuaConversion(lua_State* state) const {
+    lua_pushnumber(state, static_cast<lua_Number>(_value));
```
Comment (alexanderbock) [inline src/properties/scalar/ulongproperty.cpp:62]: ^^

File: src/properties/scalar/ushortproperty.cpp [hunk 11]
```diff
@@ -42,31 +58,8 @@ unsigned short fromLuaConversion(lua_State* state, bool& success) {
     }
 }
 
-bool toLuaConversion(lua_State* state, unsigned short value) {
-    lua_pushnumber(state, static_cast<lua_Number>(value));
-    return true;
+void UShortProperty::toLuaConversion(lua_State* state) const {
+    lua_pushnumber(state, static_cast<lua_Number>(_value));
```
Comment (alexanderbock) [inline src/properties/scalar/ushortproperty.cpp:62]: ^^

File: src/properties/stringproperty.cpp [hunk 12]
```diff
@@ -23,46 +23,36 @@
...
 
-} // namespace
+std::string StringProperty::fromLuaConversion(lua_State* state, bool& success) const {
+    success = lua_isstring(state, -1) == 1;
+    return success ? lua_tostring(state, -1) : "";
+}
 
-namespace openspace::properties {
+void StringProperty::toLuaConversion(lua_State* state) const {
+    lua_pushstring(state, _value.c_str());
```
Comment (alexanderbock) [inline src/properties/stringproperty.cpp:49]: ^^

**Review by alexanderbock — APPROVED** [review]
Looks good to me.   Due to the nature of the PR, there's a lot of "comment once" + "^^" for all other types

File: src/properties/list/doublelistproperty.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (WeirdRubberDuck) [reply src/properties/list/doublelistproperty.cpp:81]: It totally does, and this is how I implemented the toString conversion. Not applicable to the doublelistproperty though, as it's not a NumericalProperty 🙂

File: include/openspace/util/json_helper.inl [hunk 13]
```diff
@@ -22,73 +22,75 @@
...
-        "Unimplemented PropertyDelegate::toString specialization");
-    return false;
+std::string formatJson(T value) {
+    if constexpr (std::is_same_v<T, bool>) {
+        return value ? "true" : "false";
+    }
+    else if constexpr (std::is_arithmetic_v<T>) {
+        return formatJsonNumber(static_cast<double>(value));
+    }
+    else if constexpr (std::is_same_v<T, std::string>) {
```
[the remark this answers is not in the document]
Comment (WeirdRubberDuck) [reply include/openspace/util/json_helper.inl:67]: Yep, it does

**Pull request merged**